x-pack/plugins/apm/server/routes/services/get_services/get_service_transaction_stats.ts

```diff
@@ -189,6 +163,136 @@ export async function getServiceTransactionStats({
         };
       }) ?? [],
     serviceOverflowCount:
-      response.aggregations?.sample?.overflowCount.value || 0,
+      response.aggregations?.sample?.overflowCount?.value ?? 0,
+  };
+}
+
+function isDurationSummary(
+  outcome: any
+): outcome is AggregationResultOfMap<DurationSummaryAggregation, {}> {
+  const summaryAgg = outcome as AggregationResultOfMap<
+    DurationSummaryAggregation,
+    {}
+  >;
+  return !!summaryAgg.avg_duration_legacy && !!summaryAgg.avg_duration_summary;
+}
+
+function getAvgDurationValue(
+  outcomeResponse: AggregationResultOfMap<AvgLatencyAggregation, {}>
+) {
+  if (isDurationSummary(outcomeResponse)) {
+    return (
+      outcomeResponse.avg_duration_summary.result.value ??
+      outcomeResponse.avg_duration_legacy.result.value
+    );
+  }
+
+  return outcomeResponse.avg_duration.value;
+}
+
+function getDurationAggregation(summaryFieldSupportedByDocType: boolean) {
+  return {
+    avg_duration: {
+      avg: {
+        field: summaryFieldSupportedByDocType
+          ? TRANSACTION_DURATION_SUMMARY
+          : TRANSACTION_DURATION,
+      },
+    },
+  };
+}
+
+function getConditionalDurationAggregation() {
+  return {
+    avg_duration_summary: {
+      filter: { ...getDurationSummaryFilter() },
+      aggs: {
+        result: {
+          avg: {
+            field: TRANSACTION_DURATION_SUMMARY,
+          },
+        },
+      },
+    },
+    avg_duration_legacy: {
+      filter: { ...getDurationLegacyFilter() },
+      aggs: {
+        result: {
+          avg: {
+            field: TRANSACTION_DURATION_HISTOGRAM,
+          },
+        },
+      },
+    },
+  };
+}
+
+function getSearchRequest({
+  documentType,
+  rollupInterval,
+  filters,
+  maxNumServices,
+  randomSampler,
+  metrics,
+}: SearchParams) {
+  return {
+    apm: {
+      sources: [
+        {
+          documentType,
+          rollupInterval,
+        },
+      ],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: filters,
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            overflowCount: {
+              sum: {
+                field: SERVICE_OVERFLOW_COUNT,
+              },
+            },
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: maxNumServices,
+              },
+              aggs: {
+                transactionType: {
+                  terms: {
+                    field: TRANSACTION_TYPE,
+                  },
+                  aggs: {
+                    ...metrics,
+                    environments: {
+                      terms: {
+                        field: SERVICE_ENVIRONMENT,
+                      },
+                    },
+                    sample: {
+                      top_metrics: {
+                        metrics: [{ field: AGENT_NAME } as const],
```

Comment:
Q: Why do we need `as const` here?
```suggestion
                        metrics: [{ field: AGENT_NAME }],
```

Reply:
I don't know why it's creating a read-only metrics object. I just moved this code to a new function. I didn't really write this.